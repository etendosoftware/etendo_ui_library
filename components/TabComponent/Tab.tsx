{/* Imports */}
import React, {useState} from 'react';
import {
  ScrollView,
  StyleSheet,
} from 'react-native';
import TabItem from './TabItem';

{/* Type declaration */}
type Info = {
  name: string;
  route: string;
  key: string;
};

export type TabsProps = {
  data: Array<any>;
  onChangeSelected: any;
};

{/* Styles */}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'row',
  },
  tabs: {
    marginTop: 50,
    paddingHorizontal: 20,
    height: 45,
    backgroundColor: '#eff1f8',
    borderBottomWidth: 3,
    borderBottomColor: '#161f56',
    justifyContent: 'center',
    alignItems: 'center',
    borderTopWidth: 4,
    borderTopColor: 'transparent',
  },
  tabsActive: {
    backgroundColor: '#161f56',
    borderTopWidth: 4,
    borderTopColor: '#fad614',
  },
});

{/* Tab component */}
export const Tab = ({ data, onChangeSelected }: TabsProps) => {
  {/* Variable to handle the state of the tabs */}
  const [toggleItem, setToggleItem] = useState(0);

  return (
      <ScrollView horizontal={true}>
        {/* Mapping of each item */}
        {data.map((item: any, index: number) => {

          {/* TabItem rendering */}
          return (
            <TabItem
              style={toggleItem === index
                ? [styles.tabs, styles.tabsActive]
                : styles.tabs}
              item={item}
              index={index}
              onPress={() => {
                setToggleItem(index); {/* Allows to detect the selected item */}
                onChangeSelected({key: item.key, route: item.route})
              }}
              key={item.key}
              toggleItem={toggleItem} />
          );
        })}
      </ScrollView>
  );
};
