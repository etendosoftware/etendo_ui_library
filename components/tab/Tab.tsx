/* Imports */
import React, {useState} from 'react';
import {ScrollView, useWindowDimensions, View} from 'react-native';

import TabItem from './TabItem';
import {TabProps, Info, TabStyleType} from './Tab.types';
import {TabStyleVariant} from './Tab.styles';
import {GREY_BLUE_30, BLUE} from '../../styles/colors';
import Link from 'next/link';
import {useRouter} from 'next/router';

/* Styles - This function allows you to obtain the styles of the Tab component */

const getStyle = (style: TabStyleType, item: Info, pathname: any) => {
  /* Returns the style chosen by the programmer */
  if (item.route === pathname) {
    return [TabStyleVariant[style].tabs, TabStyleVariant[style].tabsActive];
  } else {
    return TabStyleVariant[style].tabs;
  }
};

/* Tab component */
const Tab = ({data, onChangeSelected, style, typeSizeText}: TabProps) => {
  /* Variable to handle the state of the tabs */
  const [toggleItem, setToggleItem] = useState<number>(-1);
  const router = useRouter();
  const {pathname} = router;

  return (
    <View style={TabStyleVariant.primary.container}>
      <ScrollView horizontal={true} showsHorizontalScrollIndicator={false}>
        {data.map((item: Info, index: number) => {
          return (
            <TabItem
              style={getStyle(style, item, pathname)}
              styleText={style}
              sizeText={typeSizeText}
              item={item}
              index={index}
              onPress={() => {
                /* Allows to detect the selected item */
                setToggleItem(index);
                onChangeSelected({
                  name: item.name,
                  route: item.route,
                  key: item.key,
                });
              }}
              key={item.key}
              toggleItem={toggleItem}
            />
          );
        })}
      </ScrollView>
    </View>
  );
};

export default Tab;
