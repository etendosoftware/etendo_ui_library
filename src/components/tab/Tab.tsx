// Imports
import React, { useEffect, useState } from 'react';
import { View, Text, Pressable, ScrollView } from 'react-native';
import {
  NEUTRAL_200,
  NEUTRAL_500,
  NEUTRAL_700,
  PRIMARY_100,
  SECONDARY_100,
  TERTIARY_50,
  TERTIARY_80,
} from '../../styles/colors';
import { TabItemType, TabProps } from './Tab.types';
import { TabStyleVariants, mapContainerViewStyle } from './Tab.styles';

const Tab = ({ data, currentIndex, onPressTab, typeStyle }: TabProps) => {
  const [hoveredTab, setHoveredTab] = useState<number | null>();
  const [currentIndexSelected, setCurrentIndexSelected] = useState<number>(0);

  useEffect(() => {
    setCurrentIndexSelected(currentIndex);
  }, [currentIndex]);

  const handleHoverIn = (index: number) => {
    setHoveredTab(index);
  };

  const handleHoverOut = () => {
    setHoveredTab(null);
  };

  const handleOnPress = (item: TabItemType, index: number) => {
    if (onPressTab) {
      onPressTab(item.route, index);
    }
    setCurrentIndexSelected(index);
  };

  const hoveredStyle = (index: number) => {
    let HoveredColor;
    if (index === hoveredTab) {
      HoveredColor = TERTIARY_80;
      return HoveredColor;
    } else {
      HoveredColor = TERTIARY_50;
      return HoveredColor;
    }
  };

  const getTabStyle = (index: number) => {
    switch (typeStyle) {
      case 'primary':
        return {
          borderBottomColor:
            index === currentIndexSelected
              ? PRIMARY_100
              : index === hoveredTab
              ? NEUTRAL_700
              : NEUTRAL_500,
          borderBottomWidth: index === currentIndexSelected ? 2 : 1,
        };
      case 'secondary':
        return {
          backgroundColor:
            index === currentIndexSelected
              ? SECONDARY_100
              : index === hoveredTab
              ? TERTIARY_50
              : NEUTRAL_200,
        };
      case 'terciary':
        return {
          backgroundColor:
            index === currentIndexSelected ? PRIMARY_100 : hoveredStyle(index),
          borderTopColor:
            index === currentIndexSelected ? SECONDARY_100 : 'transparent',
        };
    }
  };

  const [hasScroll, setHasScroll] = useState(false);

  const [containerWidth, setContainerWidth] = useState(0);
  const [mapContainer, setMapContainer] = useState(0);

  const handleContainerLayout = (event: any) => {
    const width = event.nativeEvent.layout.width;
    setContainerWidth(width);
  };
  const handleMapContainerLayout = (event: any) => {
    const width = event.nativeEvent.layout.width;
    setMapContainer(width);
  };

  useEffect(() => {
    if (mapContainer > containerWidth) {
      setHasScroll(true);
    } else {
      setHasScroll(false);
    }
  }, [containerWidth]);

  return (
    <ScrollView
      onLayout={handleContainerLayout}
      horizontal
      contentContainerStyle={[
        TabStyleVariants[typeStyle].container,
        !hasScroll && { flex: 1 },
      ]}>
      <View
        onLayout={handleMapContainerLayout}
        style={{ ...mapContainerViewStyle }}>
        {data?.map((item: TabItemType, index: number) => (
          <Pressable
            key={item.route}
            style={({ pressed }) => [
              TabStyleVariants[typeStyle].tab,
              getTabStyle(index),
            ]}
            onPress={() => handleOnPress(item, index)}
            onHoverIn={() => handleHoverIn(index)}
            onHoverOut={handleHoverOut}>
            <Text
              style={[
                index === currentIndexSelected
                  ? TabStyleVariants[typeStyle].text
                  : index === hoveredTab
                  ? TabStyleVariants[typeStyle].hoveredColor
                  : TabStyleVariants[typeStyle].textDisabled,
              ]}>
              {item.name}
            </Text>
          </Pressable>
        ))}
      </View>
    </ScrollView>
  );
};

export default Tab;
