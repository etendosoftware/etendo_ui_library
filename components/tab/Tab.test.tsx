import React from 'react';
import renderer from 'react-test-renderer';
import Tab from './Tab';

const datasetTab = [
  {name: 'Home', route: '/home', key: 'user'},
  {name: 'About', route: '/about', key: 'vehicles'},
  {name: 'Staff', route: '/staff', key: 'staff'},
  {name: 'Login', route: '/login', key: 'skills'},
  {name: 'Logout', route: '/logout', key: 'movilApp'},
];

describe('Running Test for Tab', () => {
  it('Check Tab Disabled', () => {
    const tree = renderer
      .create(
        <Tab
          data={datasetTab}
          selectedTab={1}
          setSelectedTab={() => {}}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
