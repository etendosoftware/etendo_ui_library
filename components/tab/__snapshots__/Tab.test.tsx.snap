// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running Test for Tab Check Tab Disabled 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 40,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#202452",
          "borderBottomWidth": 3,
          "borderTopWidth": 4,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
        Object {
          "backgroundColor": "#D3D6E1",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#202452",
        }
      }
    >
      Home
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#202452",
          "borderBottomWidth": 3,
          "borderTopWidth": 4,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
        Object {
          "backgroundColor": "#D3D6E1",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#202452",
        }
      }
    >
      About
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#202452",
          "borderBottomWidth": 3,
          "borderTopWidth": 4,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
        Object {
          "backgroundColor": "#D3D6E1",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#202452",
        }
      }
    >
      Staff
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#202452",
          "borderBottomWidth": 3,
          "borderTopWidth": 4,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
        Object {
          "backgroundColor": "#D3D6E1",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#202452",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#202452",
          "borderBottomWidth": 3,
          "borderTopWidth": 4,
          "flex": 1,
          "height": 40,
          "justifyContent": "center",
          "paddingHorizontal": 10,
        },
        Object {
          "backgroundColor": "#D3D6E1",
          "borderTopColor": "transparent",
        },
      ]
    }
  >
    <Text
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "#202452",
        }
      }
    >
      Logout
    </Text>
  </View>
</View>
`;
